Keep extra user_info during order migration

robincornett opened the pull request:
Fixes #8350

Proposed Changes:
1. During order migration, only removes core `user_info` from the payment meta--if anything is left, it's saved.
2. Adds a backwards compatibility check to the `setup_user_info` method in `EDD_Payment` to get the extra `user_info`.

ashleyfae commented on includes/payments/class-edd-payment.php:
```diff
@@ -3160,6 +3160,12 @@ private function setup_user_info() {
 		}
 
+		// Check for old `user_info` meta which may still exist.
+		$old_meta = $this->get_meta( 'payment_meta' );
```
I think we should be able to just use `edd_get_order_meta()` here. That's preferable because it will bypass some of the old filters and backwards compat stuff that's in `$this->get_meta()`. I don't think we need all that here.

ashleyfae reviewed (commented):
Ran a migration and this is working very well.

However, one thing I noticed is a ton of user info meta with `user_id` (note `user_id` instead of `id`).

How about we add that to our `$core_user_info` list, if it's not empty and is the same value as the normal `id`.

robincornett commented:
Updated, but I was thinking about this with Simple Shipping yesterday and left a comment there...is it worth adding a filter to the list of core `user_info` keys so that an extension could add its key(s) to the array and then not create the order meta? So Simple Shipping could add its filter with its migration process and the shipping info would not be passed on to the payment meta, since the extension is taking care of it.

I was just thinking that with this as it stands, any orders with shipping will always have the `payment_meta` created and added, even if it's not necessary. Alternatively, the extension could remove the meta when it hooks into the migration. Just thinking out loud.

ashleyfae commented:
> Updated, but I was thinking about this with Simple Shipping yesterday and left a comment there...is it worth adding a filter to the list of core `user_info` keys so that an extension could add its key(s) to the array and then not create the order meta? So Simple Shipping could add its filter with its migration process and the shipping info would not be passed on to the payment meta, since the extension is taking care of it.
> 
> I was just thinking that with this as it stands, any orders with shipping will always have the `payment_meta` created and added, even if it's not necessary. Alternatively, the extension could remove the meta when it hooks into the migration. Just thinking out loud.

Let's add a filter. :+1:

ashleyfae commented on includes/admin/upgrades/v3/class-data-migrator.php:
```diff
@@ -735,6 +734,20 @@ public static function orders( $data = null ) {
+		// Remove core keys from `user_info`.
+		$remaining_user_info = false;
+		if ( ! empty( $user_info ) ) {
+			$core_user_info      = apply_filters( 'edd_30_core_user_info', array( 'id', 'email', 'first_name', 'last_name', 'discount', 'address', 'user_id' ) );
```
Let's also add some docs to explain why you might use this filter.
